### PR TITLE
#1741, #1822 - Can't Cancel Scan, get ingest details with no associated strike causes server error

### PR DESCRIPTION
**Pull request by Fizz11**
##### Checklist

- [x] `manage.py test` passes
- [x] tests are included
- [x] documentation is changed or added

### Affected app(s)

ingest/scan
### Description of change

added scan cancel API endpoint, during development ended up fixing an ingest/scan bug as well.

### CODE REVIEW

**Comment by Fizz11** [comment]
fixes #1741, #1822

File: scale/docs/rest/v6/scan.rst [hunk 1]
```diff
@@ -632,6 +632,31 @@ Response: 200 OK
...
+This call will send a cancel job request for the job associated with the scan, as well as all queued or running ingest jobs spawned from the scan.
+recipes triggered by completed ingest jobs will not be effected. Returns the job ids of all canceled jobs. 
+
+**Example POST /v6/scans/cancel/{id} API call**
+
+Request: POST http://.../v6/scans/cancel/{id}/
+
+.. code-block:: javascript
+
+    {
```
Comment (emimaesmith) [inline scale/docs/rest/v6/scan.rst]: Are we missing anything here?

File: scale/docs/rest/v6/scan.rst [hunk 2]
```diff
@@ -632,6 +632,31 @@ Response: 200 OK
...
+    }
+
+Response: 202 Accepted
+
+.. code-block:: javascript
+
+    {
+    "[1,2,3]"
```
Comment (emimaesmith) [inline scale/docs/rest/v6/scan.rst]: Can we get a description of what these are?

File: scale/ingest/views.py [hunk 3]
```diff
@@ -339,6 +340,26 @@ def _create_v6(self, request):
...
+        """Returns the appropriate serializer based off the requests version of the REST API. """
+
+        if self.request.version == 'v6':
+            return ScanSerializerV6
+        elif self.request.version == 'v7':
+            return ScanSerializerV6
+
+    def post(self, request, scan_id):
+        try:
+            canceled_ids = Scan.objects.cancel_scan(scan_id)
```
Comment (emimaesmith) [inline scale/ingest/views.py]: I would suggest checking the request version here as well to ensure it's v6 or v7. If not, you'll want to return a 404.

File: scale/ingest/test/test_views.py [hunk 4]
```diff
@@ -912,6 +927,52 @@ def test_ingest_after_dry_run(self, mock_msg_mgr):
...
+    @patch('ingest.models.CommandMessageManager')
+    @patch('ingest.models.create_cancel_jobs_messages')
+    def test_cancel_scan_job_nothing_to_cancel(self, msg_create, mock_msg_mgr):
+        """Tests no cancel messages generated when jobs are not in a cancelable state"""
+
+        self.scan.job = job_utils.create_job()
+        self.scan.job.status = "COMPLETED"
+        self.scan.job.save()
+        self.scan.save()
+        self.ingest = ingest_test_utils.create_ingest(scan=self.scan, file_name='test3.txt',
```
Comment (emimaesmith) [inline scale/ingest/test/test_views.py]: Any reason to not set the job status as CANCELED here?

File: scale/ingest/models.py [hunk 5]
```diff
@@ -981,6 +983,36 @@ def validate_scan_v6(self, configuration):
...
+
+        # cancel all ingest jobs associated with scan
+        relevant_ingests = Ingest.objects.get_ingests_by_scan(scan_id)
+        job_ids = []
+        for ingest in relevant_ingests:
+            try:
+                if ingest.job and ingest.job.status in ['RUNNING', 'PENDING', 'QUEUED']:
+                    job_ids.append(ingest.job.id)
+            except:
+                throw()
```
Comment (emimaesmith) [inline scale/ingest/models.py]: Does this raise an actual exception here so we can track what happened?

File: scale/ingest/models.py [hunk 6]
```diff
@@ -242,7 +243,8 @@ def get_details(self, ingest_id, is_staff=False):
         ingest = ingest.defer('source_file__workspace__json_config')
         ingest = ingest.get(pk=ingest_id)
         ingest.admin_view = is_staff
-        ingest.strike.admin_view = is_staff
+        if ingest.strike:
```
Comment (emimaesmith) [inline scale/ingest/models.py]: Does scan have the same `admin_view`? If so, we should update the scan in a similar way:

```
elif ingest.scan:
    ingest.scan.admin_view = is_staff
```

Comment (Fizz11) [reply scale/ingest/models.py]: doesn't look like scans differentiate between admin/non-admin views.

File: scale/ingest/test/test_views.py (hunk 4 above)
Comment (Fizz11) [reply scale/ingest/test/test_views.py]: both completed and canceled are not 'cancelable' states, so I'm testing both at the same time here.

File: scale/ingest/models.py [hunk 7]
```diff
@@ -981,6 +983,41 @@ def validate_scan_v6(self, configuration):
...
+            try:
+                if ingest.job and ingest.job.status in ['RUNNING', 'PENDING', 'QUEUED']:
+                    job_ids.append(ingest.job.id)
+                else:
+                    if not ingest.job:
+                        logger.info("ingest %d not canceled due to no job existing", ingest.id)
+            except ex:
+                if ingest.job: 
+                    logger.warning("Job %d was not canceled.", ingest.job.id)
+                logger.exception("error in attempting to cancel ingest job")
```
Comment (emimaesmith) [inline scale/ingest/models.py]: Can we get the log/exception messages with proper capitalization?

**Pull request merged**